File: src/weaponsprocurement/core/StockPurchaseChecks.java
package weaponsprocurement.core;

import com.fs.starfarer.api.Global;
import com.fs.starfarer.api.campaign.CampaignFleetAPI;
import com.fs.starfarer.api.campaign.CargoAPI;
import com.fs.starfarer.api.campaign.SectorAPI;
import com.fs.starfarer.api.campaign.econ.MarketAPI;

final class StockPurchaseChecks {
    private StockPurchaseChecks() {
    }

    static StockPurchaseService.PurchaseResult marketContext(SectorAPI sector, MarketAPI market) {
        return sector == null || market == null
                ? StockPurchaseService.PurchaseResult.failure("No active market context.")
                : null;
    }

    static StockPurchaseService.PurchaseResult sectorContext(SectorAPI sector) {
        return sector == null
                ? StockPurchaseService.PurchaseResult.failure("No active sector context.")
                : null;
    }

    static StockPurchaseService.PurchaseResult selectedItem(StockItemType itemType, String itemId) {
        return itemId == null || itemId.isEmpty()
                ? StockPurchaseService.PurchaseResult.failure("No " + itemType.getSingularLabel().toLowerCase(java.util.Locale.US) + " selected.")
                : null;
    }

    static StockPurchaseService.PurchaseResult positiveQuantity(int quantity, String action) {
        return quantity <= 0
                ? StockPurchaseService.PurchaseResult.failure("Nothing to " + action + ".")
                : null;
    }

    static CargoAPI playerCargo(SectorAPI sector) {
        CampaignFleetAPI fleet = sector == null ? null : sector.getPlayerFleet();
        return fleet == null ? null : fleet.getCargo();
    }

    static StockPurchaseService.PurchaseResult playerCargoAvailable(CargoAPI playerCargo) {
        return playerCargo == null
                ? StockPurchaseService.PurchaseResult.failure("Player cargo is unavailable.")
                : null;
    }

    static StockPurchaseService.PurchaseResult canAfford(CargoAPI playerCargo, int totalCost) {
        return playerCargo.getCredits().get() + 0.01f < totalCost
                ? StockPurchaseService.PurchaseResult.failure("Need " + CreditFormat.creditsLong(totalCost) + " for this order.")
                : null;
    }

    static StockPurchaseService.PurchaseResult hasCargoSpace(CargoAPI playerCargo, float totalSpace) {
        return playerCargo.getSpaceLeft() + 0.01f < totalSpace
                ? StockPurchaseService.PurchaseResult.failure("Need " + Math.round(totalSpace) + " cargo space for this order.")
                : null;
    }

    static StockPurchaseService.PurchaseResult canCompletePurchase(CargoAPI playerCargo, int totalCost, float totalSpace) {
        StockPurchaseService.PurchaseResult validation = canAfford(playerCargo, totalCost);
        return validation != null ? validation : hasCargoSpace(playerCargo, totalSpace);
    }

    static StockPurchaseService.PurchaseResult canCompletePurchase(CargoAPI playerCargo, StockPurchasePlan plan) {
        return canCompletePurchase(playerCargo, plan.totalCost, plan.totalSpace);
    }

    static StockPurchaseService.PurchaseResult buyPlanAvailable(StockPurchasePlan plan, String message) {
        return plan.totalQuantity <= 0
                ? StockPurchaseService.PurchaseResult.failure(message)
                : null;
    }

    static void addCampaignMessage(String message) {
        if (Global.getSector() != null && Global.getSector().getCampaignUI() != null) {
            Global.getSector().getCampaignUI().addMessage(message);
        }
    }
}

File: src/main/kotlin/weaponsprocurement/trade/execution/StockPurchaseChecks.kt
package weaponsprocurement.trade.execution

import weaponsprocurement.stock.item.*
import weaponsprocurement.stock.market.*
import weaponsprocurement.stock.fixer.*
import weaponsprocurement.trade.plan.StockPurchasePlan
import weaponsprocurement.trade.plan.TradeMoney
import weaponsprocurement.trade.quote.CreditFormat

import com.fs.starfarer.api.Global
import com.fs.starfarer.api.campaign.CargoAPI
import com.fs.starfarer.api.campaign.SectorAPI
import com.fs.starfarer.api.campaign.econ.MarketAPI
import java.util.Locale

class StockPurchaseChecks private constructor() {
    companion object {
        @JvmStatic
        fun marketContext(sector: SectorAPI?, market: MarketAPI?): StockPurchaseService.PurchaseResult? {
            return if (sector == null || market == null) {
                StockPurchaseService.PurchaseResult.failure("No active market context.")
            } else {
                null
            }
        }

        @JvmStatic
        fun sectorContext(sector: SectorAPI?): StockPurchaseService.PurchaseResult? {
            return if (sector == null) StockPurchaseService.PurchaseResult.failure("No active sector context.") else null
        }

        @JvmStatic
        fun selectedItem(itemType: StockItemType, itemId: String?): StockPurchaseService.PurchaseResult? {
            return if (itemId.isNullOrEmpty()) {
                StockPurchaseService.PurchaseResult.failure(
                    "No " + itemType.singularLabel.lowercase(Locale.US) + " selected."
                )
            } else {
                null
            }
        }

        @JvmStatic
        fun positiveQuantity(quantity: Int, action: String): StockPurchaseService.PurchaseResult? {
            return if (quantity <= 0) StockPurchaseService.PurchaseResult.failure("Nothing to $action.") else null
        }

        @JvmStatic
        fun playerCargo(sector: SectorAPI?): CargoAPI? {
            val fleet = sector?.playerFleet
            return fleet?.cargo
        }

        @JvmStatic
        fun playerCargoAvailable(playerCargo: CargoAPI?): StockPurchaseService.PurchaseResult? {
            return if (playerCargo == null) {
                StockPurchaseService.PurchaseResult.failure("Player cargo is unavailable.")
            } else {
                null
            }
        }

        @JvmStatic
        fun canAfford(playerCargo: CargoAPI, totalCost: Long): StockPurchaseService.PurchaseResult? {
            return if (playerCargo.credits.get() + 0.01f < totalCost) {
                StockPurchaseService.PurchaseResult.failure(
                    "Need " + CreditFormat.creditsLong(totalCost) + " for this order."
                )
            } else {
                null
            }
        }

        @JvmStatic
        fun canMutateCredits(credits: Long): StockPurchaseService.PurchaseResult? {
            return if (TradeMoney.canExecuteCreditMutation(credits)) {
                null
            } else {
                StockPurchaseService.PurchaseResult.failure("Order value is too large.")
            }
        }

        @JvmStatic
        fun hasCargoSpace(playerCargo: CargoAPI, totalSpace: Float): StockPurchaseService.PurchaseResult? {
            return if (playerCargo.spaceLeft + 0.01f < totalSpace) {
                StockPurchaseService.PurchaseResult.failure(
                    "Need " + Math.round(totalSpace) + " cargo space for this order."
                )
            } else {
                null
            }
        }

        @JvmStatic
        fun canCompletePurchase(
            playerCargo: CargoAPI,
            totalCost: Long,
            totalSpace: Float,
        ): StockPurchaseService.PurchaseResult? {
            var validation = canMutateCredits(totalCost)
            if (validation != null) return validation
            validation = canAfford(playerCargo, totalCost)
            return validation ?: hasCargoSpace(playerCargo, totalSpace)
        }

        @JvmStatic
        fun canCompletePurchase(playerCargo: CargoAPI, plan: StockPurchasePlan): StockPurchaseService.PurchaseResult? {
            return canCompletePurchase(playerCargo, plan.totalCost, plan.totalSpace)
        }

        @JvmStatic
        fun buyPlanAvailable(plan: StockPurchasePlan, message: String): StockPurchaseService.PurchaseResult? {
            return if (plan.totalQuantity <= 0) StockPurchaseService.PurchaseResult.failure(message) else null
        }

        @JvmStatic
        fun addCampaignMessage(message: String?) {
            val ui = Global.getSector()?.campaignUI
            if (ui != null) ui.addMessage(message)
        }
    }
}
